File: rxhttp/src/main/java/com/senierr/http/builder/RequestBodyBuilder.java
package com.senierr.http.builder;

import java.io.File;
import java.net.FileNameMap;
import java.net.URLConnection;
import java.util.LinkedHashMap;

import io.reactivex.annotations.NonNull;
import io.reactivex.annotations.Nullable;
import okhttp3.FormBody;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;

/**
 * Http请求体构建器
 *
 * @author zhouchunjie
 * @date 2018/8/29
 */
public final class RequestBodyBuilder {

    private static final String MEDIA_TYPE_PLAIN = "text/plain; charset=utf-8";
    private static final String MEDIA_TYPE_XML = "text/xml; charset=utf-8";
    private static final String MEDIA_TYPE_JSON = "application/json; charset=utf-8";
    private static final String MEDIA_TYPE_STREAM = "application/octet-stream";

    private @Nullable RequestBody requestBody;

    private @Nullable MediaType mediaType;
    private @Nullable String stringContent;
    private @Nullable byte[] bytes;
    private @Nullable File file;

    private boolean isMultipart = false;
    private @NonNull LinkedHashMap<String, File> fileParams = new LinkedHashMap<>();
    private @NonNull LinkedHashMap<String, String> stringParams = new LinkedHashMap<>();

    public void setRequestBody(@NonNull RequestBody requestBody) {
        this.requestBody = requestBody;
    }

    public void setRequestBody4Text(@NonNull String textStr) {
        stringContent = textStr;
        mediaType = MediaType.parse(MEDIA_TYPE_PLAIN);
    }

    public void setRequestBody4JSon(@NonNull String jsonStr) {
        stringContent = jsonStr;
        mediaType = MediaType.parse(MEDIA_TYPE_JSON);
    }

    public void setRequestBody4Xml(@NonNull String xmlStr) {
        stringContent = xmlStr;
        mediaType = MediaType.parse(MEDIA_TYPE_XML);
    }

    public void setRequestBody4Byte(@NonNull byte[] bytes) {
        this.bytes = bytes;
        mediaType = MediaType.parse(MEDIA_TYPE_STREAM);
    }

    public void setRequestBody4File(@NonNull File file) {
        this.file = file;
        mediaType = guessMimeType(file.getPath());
    }

    public void isMultipart(boolean isMultipart) {
        this.isMultipart = isMultipart;
    }

    public void addRequestParam(@NonNull String key, @NonNull String value) {
        stringParams.put(key, value);
    }

    public void addRequestStringParams(@NonNull LinkedHashMap<String, String> params) {
        for (String key: params.keySet()) {
            stringParams.put(key, params.get(key));
        }
    }

    public void addRequestParam(@NonNull String key, @NonNull File file) {
        fileParams.put(key, file);
    }

    public void addRequestFileParams(@NonNull LinkedHashMap<String, File> params) {
        for (String key: params.keySet()) {
            fileParams.put(key, params.get(key));
        }
    }

    public @Nullable RequestBody build() {
        if (requestBody != null) {
            // 自定义
            return requestBody;
        } else if (stringContent != null && mediaType != null) {
            // 字符串
            return RequestBody.create(mediaType, stringContent);
        } else if (bytes != null && mediaType != null) {
            // 字节数组
            return RequestBody.create(mediaType, bytes);
        } else if (file != null && mediaType != null) {
            // 文件
            return RequestBody.create(mediaType, file);
        } else if (!fileParams.isEmpty()) {
            // 分片提交
            MultipartBody.Builder multipartBodybuilder = new MultipartBody.Builder().setType(MultipartBody.FORM);
            for (String key: fileParams.keySet()) {
                File value = fileParams.get(key);
                if (value != null) {
                    RequestBody fileBody = RequestBody.create(guessMimeType(value.getPath()), value);
                    multipartBodybuilder.addFormDataPart(key, value.getName(), fileBody);
                }
            }
            if (!stringParams.isEmpty()) {
                for (String key: stringParams.keySet()) {
                    String value = stringParams.get(key);
                    if (value != null) {
                        multipartBodybuilder.addFormDataPart(key, value);
                    }
                }
            }
            return multipartBodybuilder.build();
        } else if (!stringParams.isEmpty()) {
            if (isMultipart) {
                // 强制分片
                MultipartBody.Builder multipartBodybuilder = new MultipartBody.Builder().setType(MultipartBody.FORM);
                for (String key: stringParams.keySet()) {
                    String value = stringParams.get(key);
                    if (value != null) {
                        multipartBodybuilder.addFormDataPart(key, value);
                    }
                }
                return multipartBodybuilder.build();
            } else {
                // 默认表单
                FormBody.Builder bodyBuilder = new FormBody.Builder();
                for (String key : stringParams.keySet()) {
                    String value = stringParams.get(key);
                    if (value != null) {
                        bodyBuilder.add(key, value);
                    }
                }
                return bodyBuilder.build();
            }
        } else {
            return null;
        }
    }

    private @Nullable MediaType guessMimeType(@NonNull String path) {
        FileNameMap fileNameMap = URLConnection.getFileNameMap();
        path = path.replace("#", "");   //解决文件名中含有#号异常的问题
        String contentType = fileNameMap.getContentTypeFor(path);
        if (contentType == null) {
            contentType = MEDIA_TYPE_STREAM;
        }
        return MediaType.parse(contentType);
    }
}

File: rxhttp/src/main/java/com/senierr/http/builder/RequestBodyBuilder.kt
package com.senierr.http.builder

import okhttp3.FormBody
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import java.io.File
import java.net.URLConnection
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.RequestBody.Companion.asRequestBody

/**
 * Http请求体构建器
 *
 * @author zhouchunjie
 * @date 2018/8/29
 */
class RequestBodyBuilder : Builder<RequestBody?> {

    companion object {
        private const val MEDIA_TYPE_PLAIN = "text/plain; charset=utf-8"
        private const val MEDIA_TYPE_XML = "text/xml; charset=utf-8"
        private const val MEDIA_TYPE_JSON = "application/json; charset=utf-8"
        private const val MEDIA_TYPE_STREAM = "application/octet-stream"
    }

    private var requestBody: RequestBody? = null

    private var mediaType: MediaType? = null
    private var stringContent: String? = null
    private var bytes: ByteArray? = null
    private var file: File? = null

    private var isMultipart = false
    private val fileParams = LinkedHashMap<String, File>()
    private val stringParams = LinkedHashMap<String, String>()

    fun setRequestBody(requestBody: RequestBody) {
        this.requestBody = requestBody
    }

    fun setRequestBody4Text(textStr: String) {
        stringContent = textStr
        mediaType = MEDIA_TYPE_PLAIN.toMediaTypeOrNull()
    }

    fun setRequestBody4JSon(jsonStr: String) {
        stringContent = jsonStr
        mediaType = MEDIA_TYPE_JSON.toMediaTypeOrNull()
    }

    fun setRequestBody4Xml(xmlStr: String) {
        stringContent = xmlStr
        mediaType = MEDIA_TYPE_XML.toMediaTypeOrNull()
    }

    fun setRequestBody4Byte(bytes: ByteArray) {
        this.bytes = bytes
        mediaType = MEDIA_TYPE_STREAM.toMediaTypeOrNull()
    }

    fun setRequestBody4File(file: File) {
        this.file = file
        mediaType = guessMimeType(file.path)
    }

    fun isMultipart(isMultipart: Boolean) {
        this.isMultipart = isMultipart
    }

    fun addRequestParam(key: String, value: String) {
        stringParams[key] = value
    }

    fun addRequestStringParams(params: LinkedHashMap<String, String>) {
        stringParams.putAll(params)
    }

    fun addRequestParam(key: String, file: File) {
        fileParams[key] = file
    }

    fun addRequestFileParams(params: LinkedHashMap<String, File>) {
        fileParams.putAll(params)
    }

    override fun build(): RequestBody? {
        if (requestBody != null) {
            // 自定义
            return requestBody
        } else if (stringContent != null && mediaType != null) {
            // 字符串
            return stringContent?.toRequestBody(mediaType)
        } else if (bytes != null && mediaType != null) {
            // 字节数组
            return bytes?.toRequestBody(mediaType)
        } else if (file != null && mediaType != null) {
            // 文件
            return file?.asRequestBody(mediaType)
        } else if (fileParams.isNotEmpty()) {
            // 分片提交
            val multipartBodybuilder = MultipartBody.Builder().setType(MultipartBody.FORM)
            for (key in fileParams.keys) {
                val value = fileParams[key]
                if (value != null) {
                    val fileBody = value.asRequestBody(guessMimeType(value.path))
                    multipartBodybuilder.addFormDataPart(key, value.name, fileBody)
                }
            }
            if (stringParams.isNotEmpty()) {
                for (key in stringParams.keys) {
                    val value = stringParams[key]
                    if (value != null) {
                        multipartBodybuilder.addFormDataPart(key, value)
                    }
                }
            }
            return multipartBodybuilder.build()
        } else if (stringParams.isNotEmpty()) {
            if (isMultipart) {
                // 强制分片
                val multipartBodybuilder = MultipartBody.Builder().setType(MultipartBody.FORM)
                for (key in stringParams.keys) {
                    val value = stringParams[key]
                    if (value != null) {
                        multipartBodybuilder.addFormDataPart(key, value)
                    }
                }
                return multipartBodybuilder.build()
            } else {
                // 默认表单
                val bodyBuilder = FormBody.Builder()
                for (key in stringParams.keys) {
                    val value = stringParams[key]
                    if (value != null) {
                        bodyBuilder.add(key, value)
                    }
                }
                return bodyBuilder.build()
            }
        } else {
            return null
        }
    }

    private fun guessMimeType(path: String): MediaType? {
        val result = path.replace("#", "")   //解决文件名中含有#号异常的问题
        val fileNameMap = URLConnection.getFileNameMap()
        var contentType = fileNameMap.getContentTypeFor(result)
        if (contentType == null) {
            contentType = MEDIA_TYPE_STREAM
        }
        return contentType.toMediaTypeOrNull()
    }
}